File: src/basango/services/open_graph.py
import logging
from dataclasses import dataclass
from typing import Optional

import trafilatura

from basango.core.config import ClientConfig
from basango.services.http_client import HttpClient
from basango.services.user_agents import UserAgentProvider


@dataclass
class OpenGraphObject:
    title: Optional[str] = None
    description: Optional[str] = None
    image: Optional[str] = None
    url: Optional[str] = None


class OpenGraphProvider:
    def __init__(
        self, user_agent_provider: UserAgentProvider = UserAgentProvider(rotate=False)
    ) -> None:
        self._user_agent = user_agent_provider.og()
        self._http_client = HttpClient(
            client_config=ClientConfig(),
            default_headers={"User-Agent": self._user_agent},
        )

    def consume_url(self, url: str) -> OpenGraphObject | None:
        try:
            logging.info(f"[OpenGraphProvider] Consuming url: {url}")
            html = self._http_client.get(url).text
            return self.consume_html(html, url)
        except Exception as e:
            logging.exception(f"[OpenGraphProvider] Failed to consume url: {e}")
            return None

    @classmethod
    def consume_html(
        cls, html: str, url: Optional[str] = None
    ) -> OpenGraphObject | None:
        try:
            meta = trafilatura.extract_metadata(html, default_url=url)
            if not meta:
                return None
            return OpenGraphObject(
                title=meta.title or None,
                description=meta.description or None,
                image=meta.image or None,
                url=url,
            )
        except Exception as e:
            logging.error(f"[OpenGraphProvider] Failed to extract metadata: {e}")
            return None

File: apps/crawler/src/http/open-graph.ts
import { config } from "@basango/domain/config";
import { DEFAULT_OPEN_GRAPH_USER_AGENT } from "@basango/domain/constants";
import { ArticleMetadata } from "@basango/domain/models";
import { parse } from "node-html-parser";

import { SyncHttpClient } from "#crawler/http/http-client";
import { UserAgents } from "#crawler/http/user-agent";
import { createAbsoluteUrl } from "#crawler/utils";

/**
 * Picks the first non-empty value from the provided array.
 * @param values - An array of string values
 */
const pick = (values: Array<string | null | undefined>): string | undefined => {
  for (const value of values) {
    if (value && value.trim().length > 0) {
      return value.trim();
    }
  }
  return undefined;
};

/**
 * Extracts the content of a meta tag given its property or name.
 * @param root - The root HTML element
 * @param property - The property or name of the meta tag to extract
 */
const extract = (root: ReturnType<typeof parse>, property: string): string | null => {
  const selector = `meta[property='${property}'], meta[name='${property}']`;
  const node = root.querySelector(selector);
  if (!node) {
    return null;
  }
  return node.getAttribute("content") ?? null;
};

/**
 * OpenGraph consumer for extracting Open Graph metadata from HTML pages.
 * Uses a synchronous HTTP client to fetch the HTML content.
 *
 * @author Bernard Ngandu <bernard@devscast.tech>
 */
export class OpenGraph {
  private readonly client: Pick<SyncHttpClient, "get">;

  constructor() {
    const settings = config.crawler.fetch.client;
    const provider = new UserAgents(true, DEFAULT_OPEN_GRAPH_USER_AGENT);

    this.client = new SyncHttpClient(settings, {
      defaultHeaders: { "User-Agent": provider.og() },
      userAgentProvider: provider,
    });
  }

  /**
   * Consume a URL and extract Open Graph metadata.
   * @param url - The URL to fetch and parse
   */
  async consumeUrl(url: string): Promise<ArticleMetadata | undefined> {
    try {
      const response = await this.client.get(url);
      const html = await response.text();
      return OpenGraph.consumeHtml(html, url);
    } catch {
      return undefined;
    }
  }

  /**
   * Consume HTML content and extract Open Graph metadata.
   * @param html - HTML content as a string
   * @param url - Optional URL of the page
   */
  static consumeHtml(html: string, url: string): ArticleMetadata | undefined {
    if (!html) {
      return undefined;
    }

    const root = parse(html);
    const title = pick([extract(root, "og:title"), root.querySelector("title")?.text]);
    const description = pick([extract(root, "og:description"), extract(root, "description")]);
    const image = pick([
      extract(root, "og:image"),
      root.querySelector("img")?.getAttribute("src") ?? null,
    ]);
    const canonical = pick([
      extract(root, "og:url"),
      root.querySelector("link[rel='canonical']")?.getAttribute("href") ?? null,
      url ?? null,
    ]);
    const author = pick([extract(root, "article:author"), extract(root, "og:article:author")]);
    const publishedAt = pick([
      extract(root, "article:published_time"),
      extract(root, "og:article:published_time"),
    ]);
    const updatedAt = pick([
      extract(root, "article:modified_time"),
      extract(root, "og:article:modified_time"),
    ]);

    if (!title && !description && !image && !canonical) {
      return undefined;
    }

    return {
      author,
      description,
      image: createAbsoluteUrl(url, image ?? "") || undefined,
      publishedAt,
      title,
      updatedAt,
      url: createAbsoluteUrl(url, canonical ?? "") || undefined,
    } as ArticleMetadata;
  }
}
